FIX discrete Naive Bayes model fitting for degenerate single-class case

dpoznik opened the pull request:
closes #18974

When training multinomial naive Bayes models as part of a larger pipeline, it is possible for a degenerate case to arise, wherein there is just one class label. Prior to this change, it was possible to fit a single-class `MultinomialNB` model. One would expect that using such a model for prediction would deterministically yield the one class label. However, sometimes an `IndexError` would arise.

Ultimately, this traced to the fact that `LabelBinarizer.transform` returns an array of shape `(n_samples, 1)` when there are one _or_ two classes, but `naive_bayes._BaseDiscreteNB.fit` and `naive_bayes._BaseDiscreteNB.partial_fit` (reasonably) assumed that if the return value had one column, that meant there were exactly two classes.

With this fix, the single-class case is handled differently from the two-class case, thereby ensuring the expected behavior and eliminating the source of the `IndexError`.

On this branch, all tests run via these three commands pass:
```
pytest sklearn/naive_bayes.py
pytest doc/modules/naive_bayes.rst
pytest sklearn/tests/test_naive_bayes.py
```
The last one includes a new regression test that fails on `master`.

dpoznik committed:
Handle degenerate single-class case correctly in naive_bayes.py

dpoznik committed:
Add regression test for degenerate single-class naive Bayes

dpoznik committed:
Remove partial_fit edits

dpoznik committed:
Test three remaining _BaseDiscreteNB subclasses

dpoznik committed:
Simplify handling of single-class case

dpoznik committed:
Fix degenerage single-class case in _BaseDiscreteNB.partial_fit

dpoznik committed:
Update discrete Naive Bayes tests to cover all four types

dpoznik commented on sklearn/naive_bayes.py:
```diff
@@ -552,7 +552,7 @@ def partial_fit(self, X, y, classes=None, sample_weight=None):
         if _check_partial_fit_first_call(self, classes):
             # This is the first call to partial_fit:
             # initialize various cumulative counters
-            n_effective_classes = len(classes) if len(classes) > 1 else 2
```
There is a small behavior change here when `len(classes) == 1`.
However, the docstring implies that `len(classes)` should be 2 for binary problems:
```
        classes : array-like of shape (n_classes,), default=None
            List of all the classes that can possibly appear in the y vector.
```
Assuming this is true, then I believe the only behavior change is for the degenerate single-class case, as intended.

dpoznik commented on sklearn/tests/test_naive_bayes.py:
```diff
@@ -261,8 +263,8 @@ def test_discretenb_partial_fit(cls):
         assert_array_equal(clf1.feature_count_, clf3.feature_count_)
 
 
-@pytest.mark.parametrize('cls', [BernoulliNB, MultinomialNB, GaussianNB,
```
I am not sure why `GaussianNB` was included in a test of discrete naive Bayes classifiers, but there seemed no harm in leaving it.

dpoznik committed:
Extend test to cover more array attributes

dpoznik committed:
Fix minor docstring typos in shapes of one-axis arrays

glemaitre commented on sklearn/tests/test_naive_bayes.py:
```diff
@@ -261,8 +263,8 @@ def test_discretenb_partial_fit(cls):
         assert_array_equal(clf1.feature_count_, clf3.feature_count_)
 
 
-@pytest.mark.parametrize('cls', [BernoulliNB, MultinomialNB, GaussianNB,
```
You could rename the test `test_naive_bayes_pickle` and update the first comment. We want to check that pickling is working.

dpoznik replied:
Done in 6e51074

glemaitre commented on sklearn/tests/test_naive_bayes.py:
```diff
@@ -284,8 +286,8 @@ def test_discretenb_pickle(cls):
 
 
-@pytest.mark.parametrize('cls', [BernoulliNB, MultinomialNB, GaussianNB,
-                                 CategoricalNB])
```
Maybe renaming the `discrete` part would be good since we have the `GaussianNB`

dpoznik replied:
Done in 6e51074

glemaitre commented on sklearn/tests/test_naive_bayes.py:
```diff
@@ -438,6 +445,61 @@ def test_discretenb_coef_intercept_shape(cls):
+    # Confirm that this is the case for binary problems and degenerate
+    #     one-class problems when fitting with `fit` or `partial_fit`.
+    # Non-regression test for handling degenerate one-class case.
+
```
You can add a link to the github issue

dpoznik replied:
Done in 6e51074

glemaitre commented on sklearn/tests/test_naive_bayes.py:
```diff
@@ -438,6 +445,61 @@ def test_discretenb_coef_intercept_shape(cls):
+    def test_prediction():
+        assert clf.predict(X[:1]) == y[0]
+
+    # Binary problem, fit method
```
I think that you can parametrize this part instead of repeating the test

dpoznik replied:
Thanks for your review! I may not get a chance to follow up on this point until after the holidays, but if not, I will do so shortly afterward.

dpoznik replied:
I've parametrized this test, and I agree that it's cleaner now: 313d74f.
Thanks for the suggestion, and thanks for your review!
I want to apologize in advance that I won't be able to respond again until after the holidays, but I'd be more than happy to address any additional critiques the first week of January.

glemaitre commented on sklearn/tests/test_naive_bayes.py:
```diff
@@ -438,6 +445,61 @@ def test_discretenb_coef_intercept_shape(cls):
+    X = [[1, 0, 0], [0, 1, 0], [0, 0, 1]]
+    y = [1, 1, 2]
+
+    def test_first_axis_lengths(expected_first_axis_length):
```
if needed you can put the test outside as a private check, and use the parametrization

glemaitre reviewed (commented):
A first pass. We will need to add an entry in the what's new `v1.0.rst` as a bug fix

dpoznik committed:
Merge remote-tracking branch 'upstream/master' into handle-degenerate-single-class-nb

dpoznik committed:
Rename extant tests and update whats_new doc

dpoznik commented:
> We will need to add an entry in the what's new `v1.0.rst` as a bug fix

Cool, I've drafted this in 6e51074.

dpoznik committed:
Parametrize test_discretenb_degenerate_one_class_case

glemaitre commented on doc/whats_new/v1.0.rst:
```diff
@@ -44,6 +44,14 @@ Changelog
+  classifiers (:class:`naive_bayes.BernoulliNB`,
+  :class:`naive_bayes.CategoricalNB`, :class:`naive_bayes.ComplementNB`,
+  and :class:`naive_bayes.MultinomialNB`) now correctly handle the degenerate
+  case of a single-valued `y`. :pr:`18925` by :user:`David Poznik <dpoznik>`.
```
```suggestion
  case of a single class `y`. :pr:`18925` by :user:`David Poznik <dpoznik>`.
```

glemaitre commented on sklearn/tests/test_naive_bayes.py:
```diff
@@ -438,6 +445,45 @@ def test_discretenb_coef_intercept_shape(cls):
+@pytest.mark.parametrize('num_samples,expected_first_axis_length',
+                         [(3, 2), (2, 1)])
+def test_discretenb_degenerate_one_class_case(
+        cls,
```
```suggestion
        NaiveBayes,
```

dpoznik replied:
To make this change, I think we'd also need to change line 448, so I will do this locally.

dpoznik replied:
Done in 4c5f8d2

glemaitre commented on sklearn/tests/test_naive_bayes.py:
```diff
@@ -438,6 +445,45 @@ def test_discretenb_coef_intercept_shape(cls):
+
+    X = [[1, 0, 0], [0, 1, 0], [0, 0, 1]][:num_samples]
+    y = [1, 1, 2][:num_samples]
+    clf = cls()
```
```suggestion
    clf = NaiveBayes()
```

dpoznik replied:
Will do locally along with above and line 448

dpoznik replied:
Done in 4c5f8d2

glemaitre commented on sklearn/tests/test_naive_bayes.py:
```diff
@@ -438,6 +445,45 @@ def test_discretenb_coef_intercept_shape(cls):
+        num_samples,
+        expected_first_axis_length,
+):
+    # Most array attributes of a discrete naive Bayes classifier should have
+    #     a first-axis length equal to the number of classes. Exceptions
+    #     include: ComplementNB.feature_all_, CategoricalNB.n_categories_.
+    # Confirm that this is the case for binary problems and degenerate
+    #     one-class problems when fitting with `fit` or `partial_fit`.
+    # Non-regression test for handling degenerate one-class case:
+    #     https://github.com/scikit-learn/scikit-learn/issues/18974
```
```suggestion
    # Most array attributes of a discrete naive Bayes classifier should have
    # a first-axis length equal to the number of classes. Exceptions
    # include: ComplementNB.feature_all_, CategoricalNB.n_categories_.
    # Confirm that this is the case for binary problems and degenerate
    # one-class problems when fitting with `fit` or `partial_fit`.
    # Non-regression test for handling degenerate one-class case:
    # https://github.com/scikit-learn/scikit-learn/issues/18974
```

glemaitre commented on sklearn/tests/test_naive_bayes.py:
```diff
@@ -438,6 +445,45 @@ def test_discretenb_coef_intercept_shape(cls):
+    else:
+        clf.fit(X, y)
+    assert clf.predict(X[:1]) == y[0]
+    for attr in ['classes_', 'class_count_', 'class_log_prior_',
+                 'feature_count_', 'feature_log_prob_']:
+        if hasattr(clf, attr):  # CategoricalNB has no feature_count_
+            attribute = getattr(clf, attr)
+            if isinstance(attribute, np.ndarray):
+                assert attribute.shape[0] == expected_first_axis_length
+            else:  # CategoricalNB.feature_log_prob_ is a list
+                for element in attribute:
+                    assert element.shape[0] == expected_first_axis_length
```
```suggestion

    # Checks that attributes are the proper shape
    attributes = [
        'classes_',
        'class_count_',
        'class_log_prior_',
        'feature_count_',
        'feature_log_prob_',
    ]
    for attribute_name in attributes:
        attribute = getattr(clf, attribute_name, None)
        if attribute is None:
            # CategoricalNB has no feature_count_ attribute
            continue
        if isinstance(attribute, np.ndarray):
            assert attribute.shape[0] == expected_first_axis_length
        else:
            # CategoricalNB.feature_log_prob_ is a list
            for element in attribute:
                assert element.shape[0] == expected_first_axis_length
```

glemaitre commented on .gitignore:
```diff
@@ -55,6 +55,7 @@ reuters/
 benchmarks/bench_covertype_data/
 
 *.prefs
+.project
```
Usually we don't add unrelated change. If you could revert this one.

dpoznik replied:
Done in 4c5f8d2

glemaitre reviewed (approved):
Apart of some style changes, LGTM.

dpoznik committed:
Apply suggestions from code review

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

dpoznik committed:
Tweak some variable names and comments

dpoznik committed:
Tweak wording of comment and summary

dpoznik committed:
Simplify test parametrization

dpoznik committed:
Change parameter name in test function

dpoznik committed:
Merge remote-tracking branch 'upstream/master' into handle-degenerate-single-class-nb

glemaitre committed:
Merge branch 'master' into handle-degenerate-single-class-nb

dpoznik commented on sklearn/tests/test_naive_bayes.py:
```diff
@@ -214,15 +214,17 @@ def test_discretenb_deprecated_coef_intercept(cls):
 
-@pytest.mark.parametrize("cls", [MultinomialNB, BernoulliNB, CategoricalNB])
+@pytest.mark.parametrize("cls", [MultinomialNB, ComplementNB, BernoulliNB,
+                                 CategoricalNB])
```
@glemaitre, there are 2 tests that run on all five naive Bayes classes and 8 more than run on the four discrete naive Bayes classes. I thought it might be nice to simplify the code by defining lists up at the top:
```
DISCRETE_NAIVE_BAYES_CLASSES = [BernoulliNB, CategoricalNB, ComplementNB, MultinomialNB]
ALL_NAIVE_BAYES_CLASSES = DISCRETE_NAIVE_BAYES_CLASSES + [GaussianNB]
```
and then using these in the `@pytest.mark.parametrize('class', ...)` calls. I'd be happy to implement, but I wanted to check whether constructing an explicit list for each test is preferred?

dpoznik replied:
To be explicit, this is what I had in mind: https://github.com/dpoznik/scikit-learn/commit/2c548154420feab14fb322d6076a7bef7c8b7c35

glemaitre replied:
This would be fine with me. We tend to do that in some tests.

dpoznik replied:
Cool, thanks for confirming.

dpoznik committed:
Add class lists to simplify test parametrization

dpoznik committed:
Replace "cls" with "DiscreteNaiveBayes" or "NaiveBayes"

glemaitre commented:
@thomasjpfan @lorentzenchr Would you like to have a second look

dpoznik committed:
Merge pull request #1 from dpoznik/handle-degenerate-single-class-nb-list-classes

Centralize lists of naive Bayes classes to streamline test parameterization

dpoznik committed:
Merge remote-tracking branch 'upstream/master' into handle-degenerate-single-class-nb

lorentzenchr commented on sklearn/tests/test_naive_bayes.py:
```diff
@@ -270,12 +273,11 @@ def test_discretenb_partial_fit(cls):
-                                 CategoricalNB])
-def test_discretenb_pickle(cls):
-    # Test picklability of discrete naive Bayes classifiers
+@pytest.mark.parametrize('NaiveBayes', ALL_NAIVE_BAYES_CLASSES)
+def test_naive_bayes_pickle(NaiveBayes):
+    # Test picklability of naive Bayes classifiers
```
I think we can remove this test as pickability is tested in `check_estimators_pickle` (see utils/estimator_checks.py) in `tests/test_common.py`.
I verified it by running `pytest sklearn/tests/test_common.py -v -k BernoulliNB` and so on.

dpoznik replied:
Sounds good. Done in a43f8e520.

lorentzenchr reviewed (commented):
@dpoznik Thanks for fixing this. You need to resolve merge conflicts. Beware that we've changed the default branch from **master** to **main**. Let me/us know if you need help.

dpoznik committed:
Merge remote-tracking branch 'upstream/main' into handle-degenerate-single-class-nb

dpoznik committed:
Rename variable: n_effective_classes -> n_classes

dpoznik committed:
Remove test_naive_bayes_pickle

As per @lorentzenchr in code review: "Pickability is tested in
`check_estimators_pickle` (see `utils/estimator_checks.py`)
in `tests/test_common.py`. I verified it by running
`pytest sklearn/tests/test_common.py -v -k BernoulliNB` and so on."

dpoznik committed:
Remove unused variables

dpoznik commented on sklearn/naive_bayes.py:
```diff
@@ -569,7 +569,7 @@ def partial_fit(self, X, y, classes=None, sample_weight=None):
             # This is the first call to partial_fit:
             # initialize various cumulative counters
-            n_effective_classes = len(classes) if len(classes) > 1 else 2
+            n_effective_classes = len(classes)
```
Agreed. Done in 945c6328d.

dpoznik commented:
> Thanks for fixing this.

No problem; it was fun :)
Thanks for reviewing!

> You need to resolve merge conflicts.

This should be all set with 9bbfb22e1.

> Beware that we've changed the default branch from **master** to **main**.

Cool; thanks for the heads-up!

lorentzenchr reviewed (approved):
LGTM

lorentzenchr commented:
@glemaitre Before merging, could you shortly skim over the last changes - just to be sure:smirk:

glemaitre commented:
Thanks @dpoznik Everything seems good. Merging.
